File: problem_017.hs
import Data.List

onesWord :: Int -> String
onesWord 1 = "one"
onesWord 2 = "two"
onesWord 3 = "three"
onesWord 4 = "four"
onesWord 5 = "five"
onesWord 6 = "six"
onesWord 7 = "seven"
onesWord 8 = "eight"
onesWord 9 = "nine"
onesWord _ = ""

teensWord :: Int -> String
teensWord 10 = "ten"
teensWord 11 = "eleven"
teensWord 12 = "twelve"
teensWord 13 = "thirteen"
teensWord 14 = "fourteen"
teensWord 15 = "fifteen"
teensWord 16 = "sixteen"
teensWord 17 = "seventeen"
teensWord 18 = "eighteen"
teensWord 19 = "nineteen"
teensWord _ = ""

tensWord :: Int -> String
tensWord 2 = "twenty"
tensWord 3 = "thirty"
tensWord 4 = "forty"
tensWord 5 = "fifty"
tensWord 6 = "sixty"
tensWord 7 = "seventy"
tensWord 8 = "eighty"
tensWord 9 = "ninety"
tensWord _ = ""

digitize :: Int -> Int -> [Int]
digitize base x = reverse $ unfoldr expand x
  where
    expand 0 = Nothing
    expand n = Just (n `mod` base, n `div` base)

getDigit :: [Int] -> Int -> Int
getDigit digits digit
  | null ds = 0
  | otherwise = head ds
  where ds = take 1 $ drop digit $ reverse digits

whereNotNull :: [[a]] -> [[a]]
whereNotNull = filter (not . null)

intToWords :: Int -> [String]
intToWords x
  | x < 10 = [onesWord x]
  | x < 20 = [teensWord x]
  | x < 100 = tensWord tens : (whereNotNull [onesWord ones])
  | x < 1000 && x `mod` 100 /= 0 = [onesWord hundreds, "hundred", "and"] ++ (intToWords $ x `mod` 100)
  | x < 1000 = [onesWord hundreds, "hundred"]
  | x < 10000 = [onesWord thousands, "thousand"] ++ (whereNotNull $ intToWords (x `mod` 1000))
  where
    digits = digitize 10 x
    digit = getDigit digits
    ones = digit 0
    tens = digit 1
    hundreds = digit 2
    thousands = digit 3

main :: IO()
main = putStrLn . show . length . concat . concat $ map intToWords [1..1000]

File: java/problem_017.java
import java.util.*;
import java.lang.*;
import java.io.*;

public class problem_017
{
	static String getWordUnderTwenty(int numberUnderTwenty)
	{
		switch (numberUnderTwenty)
		{
			case 1: return "one";
			case 2: return "two";
			case 3: return "three";
			case 4: return "four";
			case 5: return "five";
			case 6: return "six";
			case 7: return "seven";
			case 8: return "eight";
			case 9: return "nine";
			case 10: return "ten";
			case 11: return "eleven";
			case 12: return "twelve";
			case 13: return "thirteen";
			case 14: return "fourteen";
			case 15: return "fifteen";
			case 16: return "sixteen";
			case 17: return "seventeen";
			case 18: return "eighteen";
			case 19: return "nineteen";
			default: return null;
		}
	}
	
	static String getTensWord(int tens)
	{
		switch(tens)
		{
			case 2: return "twenty";
			case 3: return "thirty";
			case 4: return "forty";
			case 5: return "fifty";
			case 6: return "sixty";
			case 7: return "seventy";
			case 8: return "eighty";
			case 9: return "ninety";
			default: return null;
		}
	}
	
	static String getWords(int number)
	{
		String result = null;
		if (number < 20)
		{
			result = getWordUnderTwenty(number);
		}
		else if (number < 100)
		{
			result = getTensWord(number / 10 % 10);
			String tens = getWordUnderTwenty(number % 10);
			if (tens != null)
				result += tens;
		}
		else if (number < 1000)
		{
			result = getWordUnderTwenty(number / 100 % 10);
			result += "hundred";
			String tens = getWords(number % 100);
			if (tens != null)
				result += "and" + tens;
		}
		else if (number < 10000)
		{
			result = getWordUnderTwenty(number / 1000 % 10);
			result += "thousand";
			String hundreds = getWords(number % 1000);
			if (hundreds != null)
				result += hundreds;
		}
		return result;
	}
	
	public static void main (String[] args) throws java.lang.Exception
	{
		int length = 0;
		for (int number = 1; number <= 1000; number++)
			length += getWords(number).length();
		System.out.println(length);
	}
}
